--- a/newlib/libc/include/sys/features.h
+++ b/newlib/libc/include/sys/features.h
@@ -545,6 +545,20 @@
 
 #endif /* __CYGWIN__ */
 
+/* ESP-specific.
+ * TODO: introduce 'esp-idf' sys subdirectory, provide custom features.h there.
+ */
+#if defined(__XTENSA__) || defined(__riscv)
+
+#define _POSIX_THREADS                          1
+#define _POSIX_TIMEOUTS                         1
+#define _POSIX_TIMERS                           1
+#define _POSIX_MONOTONIC_CLOCK                  200112L
+#define _POSIX_CLOCK_SELECTION                  200112L
+#define _UNIX98_THREAD_MUTEX_ATTRIBUTES         1
+
+#endif /* ESP specific */
+
 #ifdef __cplusplus
 }
 #endif
