--- a/newlib/libc/include/sys/config.h
+++ b/newlib/libc/include/sys/config.h
@@ -193,6 +193,28 @@
 #define __DYNAMIC_REENT__
 #endif
 
+#ifdef __XTENSA__
+/* Espressif-specific: shrink the default fd buffer size */
+#define __BUFSIZ__ 128
+#ifndef __DYNAMIC_REENT__
+#define __DYNAMIC_REENT__
+#endif
+#ifndef _REENT_SMALL
+#define _REENT_SMALL
+#endif
+#endif /* __XTENSA__ */
+
+#ifdef __riscv
+/* Espressif-specific: shrink the default fd buffer size */
+#define __BUFSIZ__ 128
+#ifndef __DYNAMIC_REENT__
+#define __DYNAMIC_REENT__
+#endif
+#ifndef _REENT_SMALL
+#define _REENT_SMALL
+#endif
+#endif /* __riscv */
+
 /* This block should be kept in sync with GCC's limits.h.  The point
    of having these definitions here is to not include limits.h, which
    would pollute the user namespace, while still using types of the
