https://bugs.gentoo.org/933711

--- a/src/3rdparty/2geom/include/2geom/path.h
+++ b/src/3rdparty/2geom/include/2geom/path.h
@@ -843,7 +843,7 @@ private:
     void _unshare() {
         // Called before every mutation.
         // Ensure we have our own copy of curve data and reset cached values
-        if (!_data.unique()) {
+        if (_data.use_count() != 1) {
             _data.reset(new PathData(*_data));
             _closing_seg = static_cast<ClosingSegment*>(&_data->curves.back());
         }
