--- a/src/main.c
+++ b/src/main.c
@@ -136,6 +136,7 @@
 static void
 maybe_add_rpath_introspection_paths (void)
 {
+  extern ElfW(Dyn) _DYNAMIC[];
   ElfW (Dyn) *dyn;
   ElfW (Dyn) *rpath = NULL;
   ElfW (Dyn) *runpath = NULL;
@@ -159,7 +160,7 @@
   if ((!rpath && !runpath) || !strtab)
     return;
 
-  if (dladdr1 (_DYNAMIC, &dl_info, (void **) &link_map, RTLD_DL_LINKMAP))
+  if (dladdr (_DYNAMIC, &dl_info))
     {
       /* Sanity check */
       g_return_if_fail ((void *) _DYNAMIC == (void *) link_map->l_ld);
